dali/imgcodec/decoders/nvjpeg_lossless/nvjpeg_lossless.cc

```diff
@@ -33,114 +33,188 @@ NvJpegLosslessDecoderInstance::NvJpegLosslessDecoderInstance(
       event_(CUDAEvent::Create(device_id)) {
   DeviceGuard dg(device_id_);
   CUDA_CALL(nvjpegCreateEx(NVJPEG_BACKEND_LOSSLESS_JPEG, NULL, NULL, 0, &nvjpeg_handle_));
-  CUDA_CALL(nvjpegJpegStreamCreate(nvjpeg_handle_, &jpeg_stream_));
+  per_thread_resources_.push_back(PerThreadResources{nvjpeg_handle_});
   CUDA_CALL(nvjpegJpegStateCreate(nvjpeg_handle_, &state_));
 }
 
 NvJpegLosslessDecoderInstance::~NvJpegLosslessDecoderInstance() {
   DeviceGuard dg(device_id_);
   CUDA_CALL(cudaEventSynchronize(event_));
-  CUDA_CALL(nvjpegJpegStreamDestroy(jpeg_stream_));
+  per_thread_resources_.clear();
   CUDA_CALL(nvjpegJpegStateDestroy(state_));
   CUDA_CALL(nvjpegDestroy(nvjpeg_handle_));
 }
 
+NvJpegLosslessDecoderInstance::PerThreadResources::PerThreadResources(nvjpegHandle_t handle) {
+  CUDA_CALL(nvjpegJpegStreamCreate(handle, &jpeg_stream));
+}
+
+NvJpegLosslessDecoderInstance::PerThreadResources::PerThreadResources(PerThreadResources&& other)
+  : jpeg_stream(other.jpeg_stream) {
+  other.jpeg_stream = nullptr;
+}
+
+NvJpegLosslessDecoderInstance::PerThreadResources::~PerThreadResources() {
+  if (jpeg_stream) {
+    CUDA_CALL(nvjpegJpegStreamDestroy(jpeg_stream));
+  }
+}
+
 bool NvJpegLosslessDecoderInstance::CanDecode(DecodeContext ctx, ImageSource *in, DecodeParams opts,
                                               const ROI &roi) {
   if (opts.format != DALI_ANY_DATA && opts.format != DALI_GRAY) {
     return false;
   }
 
-  try {
-    CUDA_CALL(nvjpegJpegStreamParseHeader(nvjpeg_handle_, in->RawData<unsigned char>(), in->Size(),
-                                          jpeg_stream_));
-
-    int is_supported = 0;  // 0 means yes
-    CUDA_CALL(nvjpegDecodeBatchedSupported(nvjpeg_handle_, jpeg_stream_, &is_supported));
-    return is_supported == 0;
-  } catch (...) {
+  JpegParser jpeg_parser{};
```

Comment:
Here, we avoid calling `nvjpegJpegStreamParseHeader` which is very heavy and we need to do it later again. Instead, we simply look for the SOF-3 marker and leave the parsing to ScheduleDecode

Reply:
What happens when `ScheduleDecode` fails? Can we fallback to something else in such case?
Although `nvjpegJpegStreamParseHeader` is expensive it seems to be the only way to confirm that nvJPEG can handle provided stream.

Reply:
CanDecode can be optimistic - we do that all the time in other decoders. ScheduleDecode returns partial results and the high-level decoder will redirect the failing samples to a fallback (if there's any).

Reply:
Understood.